uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/PortletsRESTController.java

```diff
@@ -67,35 +78,91 @@ public class PortletsRESTController {
 
     @Autowired private IPortletExecutionManager portletExecutionManager;
 
+    @Autowired private IUserInstanceManager userInstanceManager;
+
+    @Autowired private FavoritesUtils favoritesUtils;
+
     private final Logger logger = LoggerFactory.getLogger(getClass());
 
+    public enum PortletPermissionType {
+        BROWSE,
+        CONFIGURE,
+        MANAGE,
+        RENDER,
+        SUBSCRIBE
+    }
+
     /**
      * Provides information about all portlets in the portlet registry. NOTE: The response is
      * governed by the <code>IPermission.PORTLET_MANAGER_xyz</code> series of permissions. The
      * actual level of permission required is based on the current lifecycle state of the portlet.
      */
-    @RequestMapping(value = "/portlets.json", method = RequestMethod.GET)
-    public ModelAndView getManageablePortlets(
-            HttpServletRequest request, HttpServletResponse response) throws Exception {
+    @GetMapping(value = "/portlets.json")
+    public ModelAndView getPortlets(HttpServletRequest request) {
+        final boolean limitByFavoriteFlag = request.getParameter(FAVORITE_FLAG) != null;
+        final boolean favoriteValueToMatch =
+                Boolean.parseBoolean(request.getParameter(FAVORITE_FLAG));
+        final String requiredPermissionTypeParameter =
+                request.getParameter(REQUIRED_PERMISSION_TYPE);
+        final PortletPermissionType requiredPermissionType =
+                (requiredPermissionTypeParameter == null)
+                        ? PortletPermissionType.MANAGE
+                        : PortletPermissionType.valueOf(
+                                requiredPermissionTypeParameter.toUpperCase());
+
+        final Set<IPortletDefinition> favorites =
+                limitByFavoriteFlag ? getFavorites(request) : Collections.emptySet();
+
         // get a list of all channels
         List<IPortletDefinition> allPortlets = portletDefinitionRegistry.getAllPortletDefinitions();
         IAuthorizationPrincipal ap = getAuthorizationPrincipal(request);
 
-        List<PortletTuple> rslt = new ArrayList<PortletTuple>();
-        for (IPortletDefinition pdef : allPortlets) {
-            if (ap.canManage(pdef.getPortletDefinitionId().getStringId())) {
-                rslt.add(new PortletTuple(pdef));
+        List<PortletTuple> results = new ArrayList<>();
+        for (IPortletDefinition portletDefinition : allPortlets) {
+            if (!limitByFavoriteFlag
+                    || favorites.contains(portletDefinition) == favoriteValueToMatch) {
+                boolean hasPermission =
+                        this.doesUserHavePermissionToViewPortlet(
+                                ap, portletDefinition, requiredPermissionType);
+                if (hasPermission) {
+                    results.add(new PortletTuple(portletDefinition));
+                }
```

Comment:
I'm wondering if this could be an opportunity to try a more functional style. Not sure if this would work, but something to experiment on:
```
    Predicate<IPortletDefinition> favoritesPredicate = p -> !limitByFavoriteFlag || favorites.contains(p) == favoriteValueToMatch;
    Predicate<IPortletDefinition> permissionsPredicate = p ->doesUserHavePermissionToViewPortlet(ap, p, requiredPermissionType);

    List<PortletTuple> results = allPortlets.stream()
                                                    .filter(favoritesPredicate.and(permissionPredicate))
                                                    .map(p -> new PortletTuple(p))
                                                    .collect(Collectors.toList());
```
??

Reply:
Hmm, I haven't done much in this style but I'll give it a shot.

Reply:
Okay, done.

Reply:
It worked?! Sweet!